libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,831 @@
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            break;
+        case FrameSource::BOOTLOADER:
+            esc.set_state(ESCState::WANT_SEND_START_FW);
+            break;
+        case FrameSource::ESC:
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+            esc.set_state(ESCState::WANT_SEND_REQ_TYPE);
+#else
+#if HAL_WITH_ESC_TELEM
+            esc.set_state(ESCState::WANT_SEND_SET_TLM_TYPE);
+#else
+            esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+#endif
+#endif
+            break;
+        }
+        break;
+
+    case ESCState::WANT_SEND_START_FW:
+        return;
+    case ESCState::WAITING_OK_FOR_START_FW:
+        if (buffer_contains_ok(length)) {
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+            esc.set_state(ESCState::WANT_SEND_REQ_TYPE);
+#else
+#if HAL_WITH_ESC_TELEM
+            esc.set_state(ESCState::WANT_SEND_SET_TLM_TYPE);
+#else
+            esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+#endif
+#endif
+        }
+        break;
+
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+    case ESCState::WANT_SEND_REQ_TYPE:
+        return;
+    case ESCState::WAITING_ESC_TYPE:
+        if (length != sizeof(u.packed_esc_type)) {
+            _message_invalid_in_state_count++;
+            return;
+        }
+        esc.type = u.packed_esc_type.msg.type;
+        esc.set_state(ESCState::WANT_SEND_REQ_SW_VER);
+        break;
+
+    case ESCState::WANT_SEND_REQ_SW_VER:
+        return;
+    case ESCState::WAITING_SW_VER:
+        if (length != sizeof(u.packed_sw_ver)) {
+            _message_invalid_in_state_count++;
+            return;
+        }
+        esc.firmware_version = u.packed_sw_ver.msg.version;
+        esc.firmware_subversion = u.packed_sw_ver.msg.subversion;
+        esc.set_state(ESCState::WANT_SEND_REQ_SN);
+        break;
+
+    case ESCState::WANT_SEND_REQ_SN:
+        return;
+    case ESCState::WAITING_SN:
+        if (length != sizeof(u.packed_sn)) {
+            _message_invalid_in_state_count++;
+            return;
+        }
+        memset(esc.serial_number, '\0', ARRAY_SIZE(esc.serial_number));
+        memcpy(esc.serial_number, u.packed_sn.msg.sn, MIN(ARRAY_SIZE(esc.serial_number), ARRAY_SIZE(u.packed_sn.msg.sn)));
+#if HAL_WITH_ESC_TELEM
+        esc.set_state(ESCState::WANT_SEND_SET_TLM_TYPE);
+#else
+        esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+#endif
+        break;
+#endif
+
+#if HAL_WITH_ESC_TELEM
+    case ESCState::WANT_SEND_SET_TLM_TYPE:
+        return;
+    case ESCState::WAITING_SET_TLM_TYPE_OK:
+        if (buffer_contains_ok(length)) {
+            esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+        }
+        break;
+#endif
+
+    case ESCState::WANT_SEND_SET_FAST_COM_LENGTH:
+        return;
+    case ESCState::WAITING_SET_FAST_COM_LENGTH_OK:
+        if (buffer_contains_ok(length)) {
+            esc.set_state(ESCState::RUNNING);
+        }
+        break;
+    case ESCState::RUNNING:

```

Comment:
This is the most common case, maybe move it to the top of the switch?
Or is it a stupid micro-optimization? This is after all the 400Hz loop that some day might become the 4KHz loop or something like that

Reply:
You'd have to look at how compilers are optimising it to know whether it is worthwhile or whether it was pointless.

I ordered it like this as there's a natural progression through the states - so being nice to the reader rather than to the compiler as it were.